src/datasources/results/ResultsDataSourceBase.ts

```diff
@@ -21,8 +21,9 @@ export abstract class ResultsDataSourceBase extends DataSourceBase<ResultsQuery>
   private toDateString = '${__to:date}';
 
   readonly globalVariableOptions = (): QueryBuilderOption[] => getVariableOptions(this);
-  readonly workspacesCache = new Map<string, Workspace>([]);
-  readonly partNumbersCache: string[] = [];
+
+  static workspacesPromise: Promise<Map<string, Workspace> | void> | null = null;
+  static partNumbersPromise: Promise<string[] | void> | null = null;
```

Comment:
await: Let's not continue with the `null` alternate for the promise, instead assign the promise directly to the `readonly` property


```suggestion
  readonly workspacesCache: Promise<Map<string, Workspace>> = this.getWorkspaces()
      .then(workspaces => {
        const workspaceMap = new Map<string, Workspace>();
        if (workspaces) {
          workspaces.forEach(workspace => workspaceMap.set(workspace.id, workspace));
        }
        return workspaceMap;
      })
      .catch(error => {
        console.error('Error in loading workspaces:', error);
        return new Map<string, Workspace>();
      });

  readonly partNumbersCache: Promise<string[]> = this.queryResultsValues(ResultsPropertiesOptions.PART_NUMBER, undefined)
      .catch(error => {
        console.error('Error in loading part numbers:', error);
        return [];
      });
```

This would make sure to keep the consumption of these caches clean

Reply:
I understand your suggestion to initialize the promise immediately rather than setting it to null. However, if we do this without using a static variable, the API calls will occur twice since both the results and steps datasources extend this base class. Based on our discussion, I have updated the implementation to use a private static promise with public getters and setters to ensure the API call happens only once and is shared across subclasses.

Reply:
Thank you for clarifying the intended approach and providing the detailed example. I’ve updated the implementation accordingly to follow this pattern for both workspaces and part numbers.